### DISCUSSION TITLE
Use Output of Task in expand

### DISCUSSION THREAD

**Discussion by pargass31** [apache/airflow#30776] [General]
Hello,

I have a task who gives a dict as output but I don't manage to use this dict in the next Task, for example, I try to do something like that:

```
output_dict= task_with_dict_as_output()
task_who_use_the_previous_output.expand(output_list['value'])
```

But it doesn't work, I tried lots of stuff to access to the Xcom variable but I don't manage to access as a dict...3
How can I access to the output as a dict?

**Comment by pankajastro** [comment]
if you want to expand on dict output of a task then it will expand actually on each element of dict 
```
    @task
    def t1():
        return {"a": 1}

    @task
    def t2(x):
        print(x)

    t2.expand(x=t1())  # expand on each element of the returned dict
```

but if you want to pass output in other tasks and access as dict then probably need

```
    @task
    def t3():
        return {"x": 1}

    @task
    def t4(x):
        print(x)  # here x is dict

    t4(t3())
```

**Comment by pargass31** [comment]
thanks for your answer, in fact the output is a dict, one value in this dict is a list and I want to use this one in expand.
So I have something like that;

```
@task
def t1():
  return {'x': 'toto', 'y': [1,2,3]}

@task
def t2(value):
   print(value)
```

`t2.expand(t1['y'])`
-> ValueError: cannot map over XCom with custom Key 'y' from <Task(_PyhtonDecoratedOperator): t1>

`t2.expand(value=t1()['y'])`
-> TypeError:'_TaskDecorator' object is not subscriptable

**Reply by pargass31** [reply]
If I add a intermediary task (specifying the type of the input) it works but it not really elegant... and I don't understand why it's not possible to use the output directly.

```
    @task
    def t1():
        return {'x': 'toto', 'y': [1, 2, 3]}

    @task
    def get_the_list(value: dict):
        return value['y']

    @task
    def t2(value):
        print(value)

    t2.expand(value=get_the_list(t1()))
```

**Comment by pargass31** [comment]
Any idea?

**Reply by potiuk** [reply]
Because we could not find a better way apparently (I have no idea for details, I just found this by looking at the changes): https://github.com/apache/airflow/pull/21930 - not sure if it can be improved - maybe @uranusjr or @asb can shed more light on it but maybe you can try to see why it cannot be done and implement it @pargass31 ?

**Reply by pargass31** [reply]
Eventually I found a workaround by returning a list in place of dict and it works. Something like that:

```python
@task
def t1():
	return [{'x': 'toto', 'y': 1}, {'x': 'toto', 'y': 2}, {'x': 'toto', 'y': 3}]

@task
def t2(value: dict):
	print(value['y'])

t2.expand(value=t1())
```

**Reply by potiuk** [reply]
good idea.
